sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -655,6 +658,9 @@ case class InSet(child: Expression, hset: Set[Any]) extends UnaryExpression with
   }
 
   @transient lazy val set: Set[Any] = child.dataType match {
+    case st: StringType if !st.isUTF8BinaryCollation =>
```

Comment:
What if it's array of collated string? Shall we use `TreeSet` with collation aware ordering?

Reply:
Yep, in that case `InSet` uses `TreeSet` to which we pass an `Ordering` instance. The last one is aware of collations via
https://github.com/apache/spark/blob/aeddc3f55e604616d683bdcc3ce7c21c95b018f6/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/PhysicalDataType.scala#L46

Let me add a few checks for nested collated strings.

Reply:
@cloud-fan I added additional checks for nested collated strings: array of strings, mixed collated strings in map of strings.